criu/files-reg.c

```diff
@@ -168,52 +170,118 @@ static int copy_chunk_from_file(int fd, int img, off_t off, size_t len)
 
 	while (len > 0) {
 		ret = sendfile(img, fd, &off, len);
-		if (ret <= 0) {
+		if (ret < 0) {
 			pr_perror("Can't send ghost to image");
 			return -1;
 		}
+		if (ret == 0) {
+			break;
+		}
 
 		len -= ret;
 	}
 
 	return 0;
 }
 
-static int copy_file_to_chunks(int fd, struct cr_img *img, size_t file_size)
+// static int copy_file_to_chunks(int fd, struct cr_img *img, size_t file_size)
+// {
+// 	GhostChunkEntry ce = GHOST_CHUNK_ENTRY__INIT;
+// 	off_t data, hole = 0;
+
+// 	while (hole < file_size) {
+// 		data = lseek(fd, hole, SEEK_DATA);
+// 		if (data < 0) {
+// 			if (errno == ENXIO)
+// 				/* No data */
+// 				break;
+// 			else if (hole == 0) {
+// 				/* No SEEK_HOLE/DATA by FS */
+// 				data = 0;
+// 				hole = file_size;
+// 			} else {
+// 				pr_perror("Can't seek file data");
+// 				return -1;
+// 			}
+// 		} else {
+// 			hole = lseek(fd, data, SEEK_HOLE);
+// 			if (hole < 0) {
+// 				pr_perror("Can't seek file hole");
+// 				return -1;
+// 			}
+// 		}
+
+// 		ce.len = hole - data;
+// 		ce.off = data;
+
+// 		if (pb_write_one(img, &ce, PB_GHOST_CHUNK))
+// 			return -1;
+
+// 		if (copy_chunk_from_file(fd, img_raw_fd(img), ce.off, ce.len))
+// 			return -1;
+// 		pr_perror("sendfile wrong: %ld+%ld", ce.off, ce.len);
+
+// 	}
+
+// 	return 0;
+// }
+
+static int copy_file_to_chunks_fiemap(int fd, struct cr_img *img, size_t file_size)
 {
 	GhostChunkEntry ce = GHOST_CHUNK_ENTRY__INIT;
-	off_t data, hole = 0;
+	struct fiemap *fiemap;
+	long long extents_size;
+	long long i = 0;
 
-	while (hole < file_size) {
-		data = lseek(fd, hole, SEEK_DATA);
-		if (data < 0) {
-			if (errno == ENXIO)
-				/* No data */
-				break;
-			else if (hole == 0) {
-				/* No SEEK_HOLE/DATA by FS */
-				data = 0;
-				hole = file_size;
-			} else {
-				pr_perror("Can't seek file data");
-				return -1;
-			}
-		} else {
-			hole = lseek(fd, data, SEEK_HOLE);
-			if (hole < 0) {
-				pr_perror("Can't seek file hole");
-				return -1;
-			}
-		}
+	if ((fiemap = (struct fiemap *)malloc(sizeof(struct fiemap))) == NULL) {
+		pr_perror("Out of memory allocating fiemap");
+		return -1;
+	}
+	memset(fiemap, 0, sizeof(struct fiemap));
+
+	fiemap->fm_start = 0;
+	fiemap->fm_length = ~0; /* Lazy */
+	fiemap->fm_flags = 0;
+	fiemap->fm_extent_count = 0;
+	fiemap->fm_mapped_extents = 0;
+
+	/* Find out how many extents there are */
+	if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0) {
+		pr_perror("fiemap ioctl() failed");
+		return -1;
+	}
+
+	/* Read in the extents */
+	extents_size = sizeof(struct fiemap_extent) * (fiemap->fm_mapped_extents);
+
+	/* Resize fiemap to allow us to read in the extents */
```

Comment:
It looks like all these code has been copy-pasted from https://github.com/ColinIanKing/debug-code/blob/master/fiemap/fiemap.c#L71

It isn't bad, but we need to reformat according with the criu coding style...

Reply:
I am sorry about the wrong coding style, I hope now it looks better.